### PR TITLE
[Go 1.12] Set up Google Cloud Logging in standard AppEngine

### PR DESCRIPTION
**Pull request by KyleJu**
A part of #1747, this PR migrates the appengine logging to Google Cloud Logging in standard AppEngine

With this change,  minor behavioral changes are introduced:

- In dev_appserver, logrus is attached to  HTTP incoming request contexts.

## Test
[See logs here](https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22gae_app%22%20resource.labels.module_id%3D%22default%22%20resource.labels.version_id%3D%22gclog-standard%22%0AlogName%3D%22projects%2Fwptdashboard-staging%2Flogs%2Frequest_log%22;timeRange=PT1H?serviceId=default&project=wptdashboard-staging)

### CODE REVIEW

File: shared/appengine.go
Comment (KyleJu) [inline shared/appengine.go:63]: use ` runtimeIdentity` instead of os

**Comment by KyleJu** [comment]
Setup looks correct and surprisingly the status code is shown correctly as well

Screen Shot 2020-09-28 at 6 16 05 PM
Screen Shot 2020-10-19 at 4 26 34 PM

**Comment by KyleJu** [comment]
See another example `api/user`; the response code is 401

Screen Shot 2020-10-19 at 4 25 59 PM

File: shared/logger.go
Comment (KyleJu) [inline shared/logger.go:98]: Remove the method and replace all instances with `r.Context()`

Comment (KyleJu) [reply shared/logger.go:98]: Double check if it affects mocks

File: webapp/web/main.go
Comment (KyleJu) [inline webapp/web/main.go:31]: Address this TODO

File: shared/logger.go
Comment (KyleJu) [inline shared/logger.go:61]: This case pollutes the log with `Context without logger`; [see](https://github.com/web-platform-tests/wpt.fyi/blob/42dd9f9bc9e9e34f1cfe346395e2981e9a93c373/shared/logger.go#L162)

File: shared/logger.go
Comment (KyleJu) [reply shared/logger.go:98]: Assign `NewNilLogger` in dev_appserver

File: shared/logger.go
Comment (KyleJu) [reply shared/logger.go:61]: Will attach logrus in dev_appserver

File: shared/logger.go
Comment (KyleJu) [inline shared/logger.go:57]: Fix this TODO

File: shared/logger.go
Comment (KyleJu) [inline shared/logger.go:103]: TODO: this should be gone after merge

**Comment by KyleJu** [comment]
Structured logging seems incorrect, as it only show the same children logs

Screen Shot 2020-10-19 at 8 08 20 PM

File: shared/logger.go
Comment (KyleJu) [inline shared/logger.go:75]: maybe change the default severity to Info, consistent with the appengine default level

**Comment by KyleJu** [comment]
There is a race condition between the construction of parent/childLogger vs AddRoute, where the former is called in [main()](https://github.com/web-platform-tests/wpt.fyi/blob/9fe7d7ea2f431ed79690a1e7775ccb7cc6a17831/webapp/web/main.go#L30) and the latter is called in [init()](https://github.com/web-platform-tests/wpt.fyi/blob/9fe7d7ea2f431ed79690a1e7775ccb7cc6a17831/webapp/web/main.go#L18)

**Comment by Hexcles** [comment]
All `init`s should happen before `main`. The execution order among multiple `init`s is less clear (either undefined or shouldn't be relied on anyway).

**Comment by KyleJu** [comment]
> All `init`s should happen before `main`. The execution order among multiple `init`s is less clear (either undefined or shouldn't be relied on anyway).

I guess race condition is not exact the right word but the problem is here - Logger is used before they are initialized; used in init() but constructed in main()

**Comment by KyleJu** [comment]
I noticed the same `{}` in [searchcache](https://github.com/web-platform-tests/wpt.fyi/pull/2190) but besides this issue, the setup looks correct
Screen Shot 2020-10-21 at 10 22 08 AM (2)

**Comment by KyleJu** [comment]
@Hexcles while I am trying to figure out [{} issue](https://github.com/web-platform-tests/wpt.fyi/pull/2208#issuecomment-713616190), this code is ready for review. Notice that I will send a follow-up CL to merge the https://github.com/web-platform-tests/wpt.fyi/pull/2190 changes.

It is easier to review commit by commit.

**Comment by KyleJu** [comment]
@Hexcles Ready for review

**Review by Hexcles — APPROVED** [review]
LGTM % nits

I've pushed a commit directly to the branch to address some minor nits. PTAL.

File: api/query/cache/service/main.go
Comment (KyleJu) [inline api/query/cache/service/main.go:162]: Two small things:

- There is a case where  *projectID !=  os.Getenv("GOOGLE_CLOUD_PROJECT"); [see here](https://github.com/web-platform-tests/wpt.fyi/pull/2208/commits/74e90b3816bca2b1b030d4ef997207e66d6bfe17#diff-29135f505c736f13d37da653c9755a6a17118cd4363bee5998e54d5a04cb615cL117)

- As a result of this change, cloudtasks will also be initialized, which isn't used anywhere in flex. It looks like the library is available in the [Flex environment](https://cloud.google.com/tasks/docs/migrating), while Task Queues is only available in Standard. (The initialization is redundant but it is safe to do so)

I will also remove the initialization `monitoredResource` in the next commit

File: shared/appengine.go
Comment (KyleJu) [inline shared/appengine.go:79]: Is there a reason to use log vs logrus here?

File: api/query/cache/service/main.go
Comment (Hexcles) [reply api/query/cache/service/main.go:149]: > There is a case where *projectID != os.Getenv("GOOGLE_CLOUD_PROJECT"); see here

Great point. I decided to clarify in the help message that `--project_id` is only used for connecting to Datastore, which is useful when debugging searchcache locally. I don't think we'd ever want sending local logs to GCP. And I've removed `monitoredResource`.

> As a result of this change, cloudtasks will also be initialized, which isn't used anywhere in flex. It looks like the library is available in the Flex environment, while Task Queues is only available in Standard. (The initialization is redundant but it is safe to do so)

I think this is not a big concern. We are just holding onto a small client that's never used. When we have more (and more expensive) clients in `shared.Clients`, we'd consider conditionally initializing them.

File: shared/appengine.go
Comment (Hexcles) [reply shared/appengine.go:79]: This is on the way out (exiting the program) and we can only log to stderr, so the logging level and formatting provided by logrus are less useful here (it'd end up as plaintext in stderr either way).

**Comment by KyleJu** [comment]
> LGTM % nits
> 
> I've pushed a commit directly to the branch to address some minor nits. PTAL.

Thanks for the nit commit! I will merge after double checking the staging :)

**Comment by Hexcles** [comment]
SGTM

Let's squash-merge this PR.

**Pull request merged**
